semantic_model_generator/admin_app/app_utils.py

```diff
@@ -0,0 +1,741 @@
+                st.error(f"Invalid default_aggregation: {e}")
+
+        sample_values_df = st.data_editor(
+            pd.DataFrame(list(measure.sample_values), columns=["Sample Values"]),
+            num_rows="dynamic",
+            key=f"{table.name}-add-measure-sample-values",
+        )
+        del measure.sample_values[:]
+        for _, row in sample_values_df.iterrows():
+            if row["Sample Values"]:
+                measure.sample_values.append(row["Sample Values"])
+
+        add_button = st.form_submit_button("Add")
+
+    if add_button:
+        table.measures.append(measure)
+        st.rerun()
+
+
+@st.experimental_dialog("Edit Time Dimension")  # type: ignore[misc]
+def edit_time_dimension(
+    table_name: str, tdim: semantic_model_pb2.TimeDimension
+) -> None:
+    """
+    Renders a dialog box to edit a time dimension.
+    """
+    key_prefix = f"{table_name}-{tdim.name}"
+    tdim.name = st.text_input("Name", tdim.name, key=f"{key_prefix}-edit-tdim-name")
+    tdim.expr = st.text_input(
+        "SQL Expression", tdim.expr, key=f"{key_prefix}-edit-tdim-expr"
+    )
+    tdim.description = st.text_area(
+        "Description",
+        tdim.description,
+        key=f"{key_prefix}-edit-tdim-description",
+    )
+    synonyms_df = st.data_editor(
+        pd.DataFrame(list(tdim.synonyms), columns=["Synonyms"]),
+        num_rows="dynamic",
+        key=f"{key_prefix}-tdim-edit-measure-synonyms",
+    )
+    del tdim.synonyms[:]
+    for _, row in synonyms_df.iterrows():
+        if row["Synonyms"]:
+            tdim.synonyms.append(row["Synonyms"])
+
+    tdim.data_type = st.text_input(
+        "Data type", tdim.data_type, key=f"{key_prefix}-edit-tdim-datatype"
+    )
+    tdim.unique = st.checkbox("Does it have unique values?", value=tdim.unique)
+    sample_values_df = st.data_editor(
+        pd.DataFrame(list(tdim.sample_values), columns=["Sample Values"]),
+        num_rows="dynamic",
+        key=f"{key_prefix}-edit-tdim-sample-values",
+    )
+    del tdim.sample_values[:]
+    for _, row in sample_values_df.iterrows():
+        if row["Sample Values"]:
+            tdim.sample_values.append(row["Sample Values"])
+
+    if st.button("Save"):
+        st.rerun()
+
+
+@st.experimental_dialog("Add Time Dimension")  # type: ignore[misc]
+def add_time_dimension(table: semantic_model_pb2.Table) -> None:
+    """
+    Renders a dialog box to add a new time dimension.
+    """
+    tdim = semantic_model_pb2.TimeDimension()
+    tdim.name = st.text_input("Name", key=f"{table.name}-add-tdim-name")
+    tdim.expr = st.text_input("SQL Expression", key=f"{table.name}-add-tdim-expr")
+    tdim.description = st.text_area(
+        "Description", key=f"{table.name}-add-tdim-description"
+    )
+    synonyms_df = st.data_editor(
+        pd.DataFrame(list(tdim.synonyms), columns=["Synonyms"]),
+        num_rows="dynamic",
+        key=f"{table.name}-add-tdim-synonyms",
+    )
+    del tdim.synonyms[:]
```

Comment:
do you need these del statements?

Reply:
I think it's needed. Also I am not gonna change the app here directly (copying from our cortex codebase); we can iterate on the app offline